code/datums/components/shelved.dm

```diff
@@ -0,0 +1,241 @@
+	to_add.pixel_y = position_details["y"]
+	to_add.appearance_flags |= PIXEL_SCALE
+	if("layer" in position_details)
+		to_add.layer = position_details["layer"]
+	used_places[placement_idx] = to_add.UID()
+	var/obj/O = parent
+	if(istype(O))
+		O.update_icon(ALL)
+
+	if(default_scale)
+		to_add.transform *= default_scale
+	if(default_rotation)
+		to_add.transform = turn(to_add.transform, default_rotation)
+
+	SEND_SIGNAL(to_add, COMSIG_SHELF_ITEM_ADDED, default_scale)
+
+/datum/component/shelver/basic_shelf
+	placement_zones = list(
+		// Bottom Shelf
+		list(1,  1, 10, 16) = list("x" = -9, "y" = -5, "layer" = BELOW_OBJ_LAYER),
+		list(11, 1, 20, 16) = list("x" = 0, "y" = -5, "layer" = BELOW_OBJ_LAYER),
+		list(21, 1, 32, 16) = list("x" = 9, "y" = -5, "layer" = BELOW_OBJ_LAYER),
+
+		// Top Shelf
+		list(1,  17, 10, 32) = list("x" = -9, "y" = 9),
+		list(11, 17, 20, 32) = list("x" = 0, "y" = 9),
+		list(21, 17, 32, 32) = list("x" = 9, "y" = 9),
+	)
+
+/datum/component/shelver/gun_rack
+	placement_zones = list(
+		list(1,  1, 10, 32) = list("x" = -8, "y" = -1),
+		list(11, 1, 20, 32) = list("x" = 0, "y" = -1),
+		list(21, 1, 32, 32) = list("x" = 8, "y" = -1),
+	)
+	default_scale = 0.80
+	default_rotation = -90
+
+/datum/component/shelved
+	var/shelf_uid
+	var/matrix/original_transform
+	var/original_layer
+	var/original_appearance_flags
+
+/datum/component/shelved/Initialize(atom/shelf)
+	if(!isobj(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/obj/O = parent
+
+	shelf_uid = shelf.UID()
+	original_transform = O.transform
+	original_layer = O.layer
+	original_appearance_flags = O.appearance_flags
+
+/datum/component/shelved/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_ITEM_PICKUP, PROC_REF(on_item_pickup))
+	var/obj/shelf = locateUID(shelf_uid)
+	if(shelf)
+		RegisterSignal(shelf, COMSIG_MOVABLE_SHOVE_IMPACT, PROC_REF(on_movable_shove_impact))
+		RegisterSignal(shelf, COMSIG_ATOM_HITBY, PROC_REF(on_atom_hitby))
+		RegisterSignal(shelf, COMSIG_OBJ_DECONSTRUCT, PROC_REF(on_shelf_deconstruct))
+
+/datum/component/shelved/proc/on_shelf_deconstruct()
+	SIGNAL_HANDLER // COMSIG_OBJ_DECONSTRUCT
+	qdel(src)
+
+/datum/component/shelved/proc/on_item_pickup(obj/item/I, mob/user)
+	SIGNAL_HANDLER // COMSIG_ITEM_PICKUP
+	qdel(src)
+
+/datum/component/shelved/UnregisterFromParent()
+	. = ..()
+	var/obj/O = parent
+	O.transform = original_transform
+	O.layer = original_layer
+	O.appearance_flags = original_appearance_flags
+	O.pixel_x = 0
+	O.pixel_y = 0
+
+	var/obj/shelf = locateUID(shelf_uid)
+	if(shelf)
+		UnregisterSignal(shelf, COMSIG_MOVABLE_SHOVE_IMPACT)
+		UnregisterSignal(shelf, COMSIG_ATOM_HITBY)
+
+	SEND_SIGNAL(shelf, COMSIG_SHELF_ITEM_REMOVED, parent.UID())
+
+/datum/component/shelved/proc/scatter()
+	var/list/clear_turfs = list()
+	var/obj/O = parent
+	for(var/turf/T in orange(1, get_turf(O)))
+		if(isfloorturf(T) && T != get_turf(O))
+			clear_turfs |= T
+
+	if(length(clear_turfs))
+		var/obj/shelf = locateUID(shelf_uid)
```

Comment:
You might want to sanity check this with an `istype()`. It's unlikely, but locateUID might sometimes return something unrelated.

Reply:
Done